runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
+
+    // ----------------------------------------
+    //  DEVICE
+    // ----------------------------------------
 
     /**
-     * @brief Partial sample with the number of shots on `wires`, returning the
-     * number of counts for each sample.
+     * @brief Produce a measurement result value for "0" (computational basis).
      *
-     * @param eigvals The pre-allocated `DataView<double, 1>`
-     * @param counts The pre-allocated `DataView<int64_t, 1>`
-     * @param wires Wires to compute samples on
-     * @param shots The number of shots
+     * Results produced by projective qubit measurements (see `Measure`) have to be of the `Result`
+     * datatype. Currently, this type is defined as `bool*`, but in principle it could be a
+     * implementation-defined type supported by additional runtime functions. For the latter case,
+     * these functions exist to produce a known measurement value ("0" or "1") and compare it
+     * against another measurement value (like those produced by `Measure`). In the absence of such
+     * functionality however, these methods don't serve much of a purpose, but should be implemented
+     * nontheless for compliance with the interface.
+     *
+     * @return `Result` A "0" measurement result.
      */
-    virtual void PartialCounts(DataView<double, 1> &eigvals, DataView<int64_t, 1> &counts,
-                               const std::vector<QubitIdType> &wires, size_t shots) = 0;
+    [[nodiscard]] virtual auto Zero() const -> Result = 0;
 
     /**
-     * @brief A general measurement method that acts on a single wire.
+     * @brief Produce a measurement result value for "1" (computational basis).
      *
-     * @param wire The wire to compute Measure on
-     * @param postselect Which basis state to postselect after a mid-circuit measurement (-1 denotes
-     no post-selection)
-
-     * @return `Result` The measurement result
+     * Like `Zero`, but with a flipped result.
+     *
+     * @return `Result` A "1" measurement result.
      */
-    virtual auto Measure(QubitIdType wire, std::optional<int32_t> postselect) -> Result = 0;
+    [[nodiscard]] virtual auto One() const -> Result = 0;
 
     /**
-     * @brief Compute the gradient of a quantum tape, that is cached using
-     * `Catalyst::Runtime::Simulator::CacheManager`, for a specific set of trainable
-     * parameters.
-     *
-     * @param gradients The vector of pre-allocated `DataView<double, 1>*`
-     * to store gradients resutls for the list of cached observables.
-     * @param trainParams The vector of trainable parameters; if none, all parameters
-     * would be assumed trainable
-     *
+     * @brief (Optional) A helper method to print the state vector of a device.
      */
-    virtual void Gradient(std::vector<DataView<double, 1>> &gradients,
-                          const std::vector<size_t> &trainParams) = 0;
+    virtual void PrintState() { RT_FAIL("PrintState is unsupported by device"); }
```

Comment:
I wish we could remove this method from the interface :smile:

Reply:
I am (see the TODO) 😅

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1680/commits/c4b5a4ca1fbf2084009facf0b4ca399413b0d6a7